File: lib/date.ts
export function dayDiff(date1: Date, date2?: Date): number {
	const date2c = date2 || new Date();

	return Math.ceil(Math.abs(date2c.getTime() - date1.getTime()) / (1000 * 3600 * 24));
}

export function today(separator = '-', yearFirst = true): string {
	const date = new Date();
	const month = date.getMonth() + 1;
	const day = date.getDate();

	const dateArr = [`${month < 10 ? '0' : ''}${month}`, `${day < 10 ? '0' : ''}${day}`];

	if (yearFirst) {
		dateArr.unshift(date.getFullYear().toString());
	} else {
		dateArr.push(date.getFullYear().toString());
	}

	return dateArr.join(separator);
}

export function isValidDate(dateYYYYMMDD: string): boolean {
	if (!/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/.test(dateYYYYMMDD)) {
		throw new Error("The date format must be 'YYYY-MM-DD'");
	}

	const convertedDate: string[] = dateYYYYMMDD.split('-');

	return /^(?=\d)(?:(?:31(?!.(?:0?[2469]|11))|(?:30|29)(?!.0?2)|29(?=.0?2.(?:(?:1[6-9]|[2-9]\d)?(?:0[48]|[2468][048]|[13579][26])|(?:16|[2468][048]|[3579][26])00)(?:\x20|$))|(?:2[0-8]|1\d|0?[1-9]))([-./])(?:1[012]|0?[1-9])\1(?:1[6-9]|[2-9]\d)?\d\d(?:(?=\x20\d)\x20|$))?(((0?[1-9]|1[012])(:[0-5]\d){0,2}(\x20[AP]M))|([01]\d|2[0-3])(:[0-5]\d){1,2})?$/.test(
		`${parseInt(convertedDate[2], 10)}-${parseInt(convertedDate[1], 10)}-${parseInt(
			convertedDate[0],
			10
		)}`
	);
}

export function dateToYYYYMMDD(date: Date, separator = '-'): string {
	const month: number = date.getMonth() + 1;
	const day: number = date.getDate();

	return `${date.getFullYear()}${separator}${month < 10 ? `0${month}` : month}${separator}${
		day < 10 ? `0${day}` : day
	}`;
}

export function createDateListFromRange(startDate: Date, endDate: Date): string[] {
	if (!isValidDate(dateToYYYYMMDD(startDate)) || !isValidDate(dateToYYYYMMDD(endDate))) {
		throw new Error('Either the start date or end date is an invalid date.');
	}

	const dateDiff = Math.floor(
		(Date.parse(endDate.toString()) - Date.parse(startDate.toString())) / 86400000
	);

	if (dateDiff < 0) {
		throw new Error('The start date is more recent than the end date.');
	}

	const endDateStr: string = dateToYYYYMMDD(endDate);
	const allDate: string[] = [];
	let currentYear: number = startDate.getFullYear();
	let currentMonth: number = startDate.getMonth() + 1;
	let currentDay: number = startDate.getDate();
	let currentDateStr = '';

	const createNewDateStr = (year: number, month: number, day: number): string =>
		`${year}-${month < 10 ? '0' : ''}${month}-${day < 10 ? '0' : ''}${day}`;

	while (endDateStr !== currentDateStr) {
		if (/[0-9]{4}-12-31/g.test(currentDateStr)) {
			currentYear += 1;
			currentMonth = 1;
			currentDay = 1;
		}

		const currentNewDateStr = createNewDateStr(currentYear, currentMonth, currentDay);

		if (isValidDate(currentNewDateStr)) {
			currentDay += 1;
			allDate.push(currentNewDateStr);
			currentDateStr = currentNewDateStr;
		} else {
			currentMonth += 1;
			currentDay = 1;
			currentDateStr = createNewDateStr(currentYear, currentMonth, currentDay);
		}
	}

	return allDate;
}

File: packages/dart/lib/src/date.dart
/// Calculates the difference between two given dates and returns the number of days.
int dayDiff(DateTime date1, [DateTime? date2]) {
  final DateTime date2c = date2 ?? DateTime.now();

  return (date2c.difference(date1).inHours / 24).ceil();
}

/// Returns today's date.
String today({String separator = '-', bool yearFirst = true}) {
  final DateTime date = DateTime.now();
  final int month = date.month;
  final int day = date.day;

  final String monthStr = '${month < 10 ? '0' : ''}$month';
  final String dayStr = '${day < 10 ? '0' : ''}$day';

  if (yearFirst) {
    return '${date.year}$separator$monthStr$separator$dayStr';
  } else {
    return '$monthStr$separator$dayStr$separator${date.year}';
  }
}

/// Checks if a given date actually exists. Check only in `YYYY-MM-DD` format.
bool isValidDate(String dateYYYYMMDD) {
  if (!RegExp(r'^[0-9]{4}-[0-9]{2}-[0-9]{2}$').hasMatch(dateYYYYMMDD)) {
    throw ArgumentError("The date format must be 'YYYY-MM-DD'");
  }

  final List<String> parts = dateYYYYMMDD.split('-');
  final int year = int.parse(parts[0]);
  final int month = int.parse(parts[1]);
  final int day = int.parse(parts[2]);

  // Support range: 1600-01-01 ~ 9999/12/31
  if (year < 16 || year > 9999) return false;
  if (month < 1 || month > 12) return false;
  if (day < 1) return false;

  final int maxDay = DateTime(year, month + 1, 0).day;
  return day <= maxDay;
}

/// Returns the date data of a Date object in the format `YYYY-MM-DD`.
String dateToYYYYMMDD(DateTime date, [String separator = '-']) {
  final int month = date.month;
  final int day = date.day;

  return '${date.year}$separator${month < 10 ? '0$month' : '$month'}$separator${day < 10 ? '0$day' : '$day'}';
}

/// Create an array list of all dates from `startDate` to `endDate` in the format `YYYY-MM-DD`.
List<String> createDateListFromRange(DateTime startDate, DateTime endDate) {
  if (!isValidDate(dateToYYYYMMDD(startDate)) ||
      !isValidDate(dateToYYYYMMDD(endDate))) {
    throw ArgumentError(
        'Either the start date or end date is an invalid date.');
  }

  final int dateDiff = endDate.difference(startDate).inDays;

  if (dateDiff < 0) {
    throw ArgumentError('The start date is more recent than the end date.');
  }

  final List<String> allDate = <String>[];
  DateTime current = DateTime(startDate.year, startDate.month, startDate.day);
  final DateTime end = DateTime(endDate.year, endDate.month, endDate.day);

  while (!current.isAfter(end)) {
    allDate.add(dateToYYYYMMDD(current));
    current = current.add(const Duration(days: 1));
  }

  return allDate;
}
